secret/scanner.go

```diff
@@ -39,18 +48,75 @@ type ScanArgs struct {
 
 func (s Scanner) Scan(args ScanArgs) types.Secret {
 	var findings []types.SecretFinding
+
+	if s.AllowList.Paths != nil && len(s.AllowList.Paths) > 0 {
+		for _, path := range s.AllowList.Paths {
+			if path.MatchString(args.FilePath) {
+				return types.Secret{
+					FilePath: args.FilePath,
+					Findings: findings,
+				}
+			}
+		}
+	}
+
 	for _, rule := range s.Rules {
 		// Check if the file path should be scanned by this rule
 		if rule.Path != nil && !rule.Path.MatchString(args.FilePath) {
 			continue
 		}
 
+		if rule.AllowList.Paths != nil && len(rule.AllowList.Paths) > 0 {
+			for _, path := range rule.AllowList.Paths {
+				if path.MatchString(args.FilePath) {
+					continue
+				}
+			}
+		}
+
 		// Detect secrets
-		locations := rule.Regex.FindAllIndex(args.Content, -1)
-		if len(locations) == 0 {
+		optionalLocations := rule.Regex.FindAllIndex(args.Content, -1)
+		if len(optionalLocations) == 0 {
 			continue
 		}
 
+		// Find allowed locations
+		allowedLocations := make([][]int, 0)
+		if rule.AllowList.Regexes != nil && len(rule.AllowList.Regexes) > 0 {
+			if s.AllowList.Regexes != nil && len(s.AllowList.Regexes) > 0 {
+				rule.AllowList.Regexes = append(rule.AllowList.Regexes, s.AllowList.Regexes...)
+			}
+			for _, regex := range rule.AllowList.Regexes {
+				allowedLocations = append(allowedLocations, regex.FindAllIndex(args.Content, -1)...)
```

Comment:
I don't want to apply allow rules to the entire file content. What if we apply allow rules to detected values only? In other words, the cost should be much smaller to apply them to `match`.

Reply:
Scanning the `match` itself will be faster but less effective.
When you scan for false positives in generic rules(not a specific secret form), many False Positives are detected by the surrounding code... I think this approach is more flexible.
It will also support cases of multiline regexs' in the allow list.
For example, if the detected "secret" is inside an `enum` declaration, we will be able to ignore it.

Reply:
I'm confused since even generic rules capture values in `match` anyway.

>For example, if the detected "secret" is inside an enum declaration, we will be able to ignore it.

Would you elaborate on it?

Reply:
Oh, my bad. I wanted to say the global allow rules should be evaluated once. Global allow rules are added here and it will be evaluated many times if multiple secrets are found.
https://github.com/aquasecurity/fanal/blob/622a930f2e4307a58235a1d57f1250582deed10d/secret/scanner.go#L82-L84